flips/20220203-capability-controllers.md

```diff
@@ -0,0 +1,341 @@
+
+```cadence
+let publicAccount = getAccount(issuerAddress)
+let countCap = publicAccount.getCapability<&{HasCount}>(/public/hasCount)
+let countRef = countCap.borrow()!
+countRef.count
+```
+
+### Private Capabilities
+
+To allow only certain accounts/resources (read) access to the `count` field on the counter, the _issuer_ needs to create a private typed link at a chosen path that points to their stored counter resource.
+
+```cadence
+AuthAccount.link<&{HasCount}>(/private/hasCount, target: /storage/counter)
+```
+
+The receiving party needs to offer a public way to receive `&{HasCount}` capabilities.
+
+```cadence
+// this would probably be defined on the same contract as `HasCount`
+pub resource interface HasCountReceiverPublic {
+   pub fun addCapability(cap: Capability<&{HasCount}>)
+}
+
+pub resource HasCountReceiver: HasCountReceiverPublic {
+
+   pub var hasCountCapability: Capability<&{HasCount}>?
+
+   init() {
+       self.hasCountCapability = nil
+   }
+
+   pub fun addCapability(cap: Capability<&{HasCount}>) {
+       self.hasCountCapability = cap
+   }
+}
+
+//...
+// this is the receiver account setup
+let hasCountReceiver <- HasCountReceiver()
+
+receivingPartyAuthAccount.save(<-hasCountReceiver, to: /storage/hasCountReceiver)
+receivingPartyAuthAccount.link<&{HasCountReceiverPublic}>(/public/hasCountReceiver,
+   target: /storage/hasCountReceive)
+```
+
+With this in place the _issuer_ can create a capability from its private link and send it to this receiver.
+
+```cadence
+let countCap = authAccount.getCapability<&{HasCount}>(/private/hasCount)
+
+let publicAccount = getAccount(receivingPartyAddress)
+let countReceiverCap = publicAccount.getCapability<&{HasCountReceiverPublic}>(/public/hasCountReceiver)
+let countReceiverRef = countCap.borrow()!
+countRef.addCapability(cap: countCap)
+```
+
+The receiving party can then use this capability when they wish.
+
+### Capability API requirements
+
+There are two requirements of the capability system that must be satisfied.
+
+- **Revocation**: Any capability must be revocable by the issuer.
+- **Redirection**: The issuer should have the ability to redirect the capability to a different (compatible) object.
+
+Revocation can be currently done by using the `unlink` function.
+
+In the public example this would mean calling `authAccount.unlink<&{HasCount}>(/public/hasCount)` which would invalidate (break) all capabilities created from this public path (both those created before unlink was called and those created after unlink was called).
+
+In the private example the call would change to `authAccount.unlink<&{HasCount}>(/private/hasCount)`. It is important to note that if the issuer wants to have the ability to revoke/redirect capabilities in a more granular way (instead of doing them all at once), the solution is to create multiple private linked paths (e.g. `/private/hasCountAlice`, `/private/hasCountBob`).
```

Comment:
```suggestion
In the public example this would mean calling `issuer.unlink<&{HasCount}>(/public/hasCount)` which would invalidate (break) all capabilities created from this public path (both those created before unlink was called and those created after unlink was called).

In the private example the call would change to `issuer.unlink<&{HasCount}>(/private/hasCount)`. It is important to note that if the issuer wants to have the ability to revoke/redirect capabilities in a more granular way (instead of doing them all at once), the solution is to create multiple private linked paths (e.g. `/private/hasCountAlice`, `/private/hasCountBob`).
```